### DISCUSSION TITLE
Plan to improve the profiler

### DISCUSSION THREAD

**Discussion by yungyuc** [solvcon/modmesh#345] [General]
It has been [discussed](https://discord.com/channels/730297880140578906/1220627986965856277) that modmesh can use enhancements to [the current prototype of the profiling systems](https://github.com/solvcon/modmesh/blob/69b7e899bcb554cf6068951f6c79048017e2b7fe/cpp/modmesh/toggle/RadixTree.hpp#L178).

Some working items are identified:

1. We should start with integrating the profiling of Python and C++.  If possible, we can try to merge the call stacks from both Python and C++ in the profiling entries.
2. To assist the development, we will visualize the profile results held in the data structures in memory.
3. Develop serialization (and deserialization) of the profiling data structures.  It allows preserving the results for future inspection (visualization in UI), and sharing among contributors.

These ideas are contributed by @tigercosmos @q40603 @kuomc and others.  An older discussion can be found at #135 .  We will need more discussions and prototypes to move it forward.

**Comment by yungyuc** [comment]
The plan derived two concrete issues #343 and #344.  They are the prerequisite for the visualization work, which can use a new issue later.

**Reply by yungyuc** [reply]
@tigercosmos @q40603 @ThreeMonth03 please share what you think here.

**Reply by tigercosmos** [reply]
Later we can work on:
1. introduce a generic JSON parser.
2. introduce the schema for profile data and even other files in the system

**Reply by yungyuc** [reply]
I guess you meant JSON schema?  Do you have any concrete examples of schema of other things?

**Reply by tigercosmos** [reply]
@yungyuc sorry, yes, I meant [JSON Schema](https://json-schema.org/), and we can use the schema to validate the data.
So far we don't have other JSON files/data, probably the configuration file in the future.

**Reply by yungyuc** [reply]
> @yungyuc sorry, yes, I meant [JSON Schema](https://json-schema.org/), and we can use the schema to validate the data. So far we don't have other JSON files/data, probably the configuration file in the future.

Thanks, that is clear.  We've discussed it in PR.  @ThreeMonth03 I am especially interested in your comments.
